src/org/exist/storage/XQueryPool.java

```diff
@@ -160,78 +155,57 @@ private synchronized void returnObject(final Source source, final CompiledXQuery
         }
     }
 
-    private Object borrowObject(final DBBroker broker, final Source source) {
-        final Source key;
-        final CompiledXQuery query;
-        synchronized (this) {
-            final int idx = getIndex(source);
-            if (idx < 0) {
-                return null;
-            }
-            key = (Source) keys[idx];
-            int validity = key.isValid(broker);
-            if (validity == Source.UNKNOWN) {
-                validity = key.isValid(source);
-            }
+    private Object borrowObject(final Source source) {
```

Comment:
Instead of needing to rebase, etc., how about if we "squash and merge"? See https://github.com/blog/2141-squash-your-commits. Looks like there are conflicts now though, so once the conflict is cleared up, this could be merged in as a single commit.